numbat/src/vm.rs

```diff
@@ -861,13 +865,60 @@ impl Vm {
                 Op::JoinString => {
                     let num_parts = self.read_u16() as usize;
                     let mut joined = String::new();
+                    let to_str = |value| match value {
+                        Value::Quantity(q) => q.to_string(),
+                        Value::Boolean(b) => b.to_string(),
+                        Value::String(s) => s,
+                        Value::DateTime(dt) => crate::datetime::to_rfc2822_save(&dt),
+                        Value::FunctionReference(r) => r.to_string(),
+                        Value::FormatSpecifiers(_) => unreachable!(),
+                    };
                     for _ in 0..num_parts {
                         let part = match self.pop() {
-                            Value::Quantity(q) => q.to_string(),
-                            Value::Boolean(b) => b.to_string(),
-                            Value::String(s) => s,
-                            Value::DateTime(dt) => crate::datetime::to_rfc2822_save(&dt),
-                            Value::FunctionReference(r) => r.to_string(),
+                            Value::FormatSpecifiers(Some(specifiers)) => match self.pop() {
+                                Value::Quantity(q) => {
+                                    let mut vars = HashMap::new();
+                                    vars.insert("value".to_string(), q.unsafe_value().to_f64());
+
+                                    let mut str =
+                                        strfmt::strfmt(&format!("{{value{}}}", specifiers), &vars)
+                                            .map_err(|err| match err {
+                                                strfmt::FmtError::Invalid(s) => {
+                                                    RuntimeError::InvalidFormatSpecifiers(s)
+                                                }
+                                                strfmt::FmtError::KeyError(_) => unreachable!(),
+                                                strfmt::FmtError::TypeError(s) => {
+                                                    RuntimeError::InvalidTypeForFormatSpecifiers(s)
+                                                }
+                                            })?;
```

Comment:
Can we please extract this into a function to avoid the duplication?

Reply:
The best I could do was extracting the function passed to `map_err`. I couldn't get the `strfmt::strfmt(...)` portion to be shared between the two branches because `vars` has a different type in each branch (in one its values are of type `f64`, in the other of type `String`. I could maybe do some generic shenanigans or, god forbid, use a macro, but doesn't seem worth it to me.